src/patchelf.cc

```diff
@@ -512,8 +516,8 @@ void ElfFile<ElfFileParamNames>::shiftFile(unsigned int extraPages, size_t start
     wri(phdr.p_offset, phdrs.at(splitIndex).p_offset - splitShift - shift);
     wri(phdr.p_paddr, phdrs.at(splitIndex).p_paddr - splitShift - shift);
     wri(phdr.p_vaddr, phdrs.at(splitIndex).p_vaddr - splitShift - shift);
-    wri(phdr.p_filesz, wri(phdr.p_memsz, splitShift + shift));
-    wri(phdr.p_flags, PF_R | PF_W);
+    wri(phdr.p_filesz, wri(phdr.p_memsz, splitShift + extraBytes));
```

Comment:
Just to check: can you post what the sections look like for you after this change? I think this is right but just wanted to make sure I'm thinking it through correctly.

Reply:
> > The Linux kernel's ELF loader (at least as of v4.18) does not accept this as a valid ELF executable
> 
> It's a weird one because I've never been able to reproduce the issue you mentioned - it seems like a certain kernel option is needed or something to make it actually reject the ELF, since it doesn't happen on my default Ubuntu install using an identical binary (sha256-matching) to what doesn't work for some others. I even deployed the latest patchelf to a few thousand users and the only feedback was from RHEL users. WSL2 too initially, but Microsoft already had a beta with the issue fixed which has now been rolled out for everyone.
> 
> Very odd, but at least you are able to reproduce it. I agree the general fix would be to create some empty space between the sections just to make sure they're on different pages since I think permissions etc work under that assumption.

I've seen the same; I'm not sure what combination of kernel version/configuration triggers it, but I too have only observed it on RHEL (and its free derivatives).

Reply:
Yeah, I should have made that more clear in the beginning. Here is the output of `readelf -l <file>` for the three cases (again, from a Python 3.10 executable that I built from source)

### Unmodified executable used as input to `patchelf`
```
Elf file type is EXEC (Executable file)
Entry point 0x4006a0
There are 9 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000400040 0x0000000000400040
                 0x00000000000001f8 0x00000000000001f8  R      0x8
  INTERP         0x0000000000000238 0x0000000000400238 0x0000000000400238
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
                 0x0000000000000948 0x0000000000000948  R E    0x200000
  LOAD           0x0000000000000d90 0x0000000000600d90 0x0000000000600d90
                 0x0000000000000294 0x0000000000000298  RW     0x200000
  DYNAMIC        0x0000000000000da0 0x0000000000600da0 0x0000000000600da0
                 0x0000000000000240 0x0000000000000240  RW     0x8
  NOTE           0x0000000000000254 0x0000000000400254 0x0000000000400254
                 0x0000000000000044 0x0000000000000044  R      0x4
  GNU_EH_FRAME   0x0000000000000828 0x0000000000400828 0x0000000000400828
                 0x000000000000003c 0x000000000000003c  R      0x4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  GNU_RELRO      0x0000000000000d90 0x0000000000600d90 0x0000000000600d90
                 0x0000000000000270 0x0000000000000270  R      0x1

 Section to Segment mapping:
  Segment Sections...
   00
   01     .interp
   02     .interp .note.ABI-tag .note.gnu.build-id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_r .rela.dyn .rela.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame
   03     .init_array .fini_array .dynamic .got .got.plt .data .bss
   04     .dynamic
   05     .note.ABI-tag .note.gnu.build-id
   06     .eh_frame_hdr
   07
   08     .init_array .fini_array .dynamic .got
```

### Output of `patchelf` master branch

This is the version that segfaults upon execution.

```
Elf file type is EXEC (Executable file)
Entry point 0x4006a0
There are 11 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x00000000003ff040 0x00000000003ff040
                 0x0000000000000268 0x0000000000000268  R      0x8
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  LOAD           0x0000000000000000 0x00000000003ff000 0x00000000003ff000
                 0x0000000000001594 0x0000000000001594  RW     0x1000
  INTERP         0x00000000000005e0 0x00000000003ff5e0 0x00000000003ff5e0
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  NOTE           0x0000000000000600 0x00000000003ff600 0x00000000003ff600
                 0x0000000000000020 0x0000000000000020  R      0x4
  NOTE           0x0000000000000620 0x00000000003ff620 0x00000000003ff620
                 0x0000000000000024 0x0000000000000024  R      0x4
  LOAD           0x0000000000001594 0x0000000000400594 0x0000000000400594
                 0x00000000000003b4 0x00000000000003b4  R E    0x1000
  GNU_EH_FRAME   0x0000000000001828 0x0000000000400828 0x0000000000400828
                 0x000000000000003c 0x000000000000003c  R      0x4
  LOAD           0x0000000000001d90 0x0000000000600d90 0x0000000000600d90
                 0x0000000000000294 0x0000000000000298  RW     0x1000
  GNU_RELRO      0x0000000000001d90 0x0000000000600d90 0x0000000000600d90
                 0x0000000000000270 0x0000000000000270  R      0x1
  DYNAMIC        0x0000000000001da0 0x0000000000600da0 0x0000000000600da0
                 0x0000000000000240 0x0000000000000240  RW     0x8

 Section to Segment mapping:
  Segment Sections...
   00
   01
   02     .dynstr .dynsym .gnu.hash .interp .note.ABI-tag .note.gnu.build-id
   03     .interp
   04     .note.ABI-tag
   05     .note.gnu.build-id
   06     .gnu.version .gnu.version_r .rela.dyn .rela.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame
   07     .eh_frame_hdr
   08     .init_array .fini_array .dynamic .got .got.plt .data .bss
   09     .init_array .fini_array .dynamic .got
   10     .dynamic
```

### Output of `patchelf` from this branch

```
Elf file type is EXEC (Executable file)
Entry point 0x4006a0
There are 11 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x00000000003ff040 0x00000000003ff040
                 0x0000000000000268 0x0000000000000268  R      0x8
  GNU_STACK      0x0000000000001000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  LOAD           0x0000000000000000 0x00000000003ff000 0x00000000003ff000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  INTERP         0x00000000000005e0 0x00000000003ff5e0 0x00000000003ff5e0
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  NOTE           0x0000000000000600 0x00000000003ff600 0x00000000003ff600
                 0x0000000000000020 0x0000000000000020  R      0x4
  NOTE           0x0000000000000620 0x00000000003ff620 0x00000000003ff620
                 0x0000000000000024 0x0000000000000024  R      0x4
  LOAD           0x0000000000001000 0x0000000000400000 0x0000000000400000
                 0x0000000000000948 0x0000000000000948  R E    0x1000
  GNU_EH_FRAME   0x0000000000001828 0x0000000000400828 0x0000000000400828
                 0x000000000000003c 0x000000000000003c  R      0x4
  LOAD           0x0000000000001d90 0x0000000000600d90 0x0000000000600d90
                 0x0000000000000294 0x0000000000000298  RW     0x1000
  GNU_RELRO      0x0000000000001d90 0x0000000000600d90 0x0000000000600d90
                 0x0000000000000270 0x0000000000000270  R      0x1
  DYNAMIC        0x0000000000001da0 0x0000000000600da0 0x0000000000600da0
                 0x0000000000000240 0x0000000000000240  RW     0x8

 Section to Segment mapping:
  Segment Sections...
   00
   01
   02     .dynstr .dynsym .gnu.hash .interp .note.ABI-tag .note.gnu.build-id
   03     .interp
   04     .note.ABI-tag
   05     .note.gnu.build-id
   06     .gnu.version .gnu.version_r .rela.dyn .rela.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame
   07     .eh_frame_hdr
   08     .init_array .fini_array .dynamic .got .got.plt .data .bss
   09     .init_array .fini_array .dynamic .got
   10     .dynamic
```